File: node/wrappers.js
/**
 * Idiomatic Node.js wrappers for the Sia SDK.
 *
 * This module provides convenience classes that make the SDK feel native
 * to Node.js developers, wrapping the low-level Reader/Writer traits with
 * standard Buffer support.
 */

import { Builder as _Builder } from "./sia_storage_ffi.js";

const DEFAULT_UPLOAD_OPTIONS = { max_inflight: 10, data_shards: 10, parity_shards: 20, progress_callback: undefined };
const DEFAULT_DOWNLOAD_OPTIONS = { max_inflight: 10, offset: 0n, length: undefined };

class BufferReader {
  #buffer;
  #offset;
  #chunkSize;

  constructor(buffer, chunkSize = 65536) {
    this.#buffer = buffer;
    this.#offset = 0;
    this.#chunkSize = chunkSize;
  }

  async read() {
    if (this.#offset >= this.#buffer.length) {
      return new Uint8Array(0);
    }
    const end = Math.min(this.#offset + this.#chunkSize, this.#buffer.length);
    const chunk = this.#buffer.subarray(this.#offset, end);
    this.#offset = end;
    return chunk;
  }
}

class BufferWriter {
  #chunks;

  constructor() {
    this.#chunks = [];
  }

  async write(data) {
    if (data.length > 0) {
      this.#chunks.push(data.slice());
    }
  }

  toBuffer() {
    return Buffer.concat(this.#chunks);
  }
}

export class Builder {
  #inner;

  /**
   * Creates a new SDK builder with the provided indexer URL.
   *
   * After creating the builder, call connected() to attempt to connect
   * using an existing app key, or request_connection() to request a new
   * connection.
   */
  constructor(indexer_url, app_meta) {
    this.#inner = new _Builder(indexer_url, app_meta);
  }

  async request_connection() {
    await this.#inner.request_connection();
    return this;
  }

  response_url() {
    return this.#inner.response_url();
  }

  async wait_for_approval() {
    await this.#inner.wait_for_approval();
    return this;
  }

  /**
   * Registers the application with the indexer using the provided mnemonic.
   * Returns an SDK instance that can be used to interact with the indexer.
   */
  async register(mnemonic) {
    return new SDK(await this.#inner.register(mnemonic));
  }

  /**
   * Attempts to connect using the provided app key.
   * Returns an SDK instance if valid, otherwise undefined.
   */
  async connected(app_key) {
    const result = await this.#inner.connected(app_key);
    return result ? new SDK(result) : undefined;
  }
}

export class SDK {
  #inner;

  /** @internal Use Builder.register() or Builder.connected() */
  constructor(inner) {
    this.#inner = inner;
  }

  /**
   * Uploads data to the Sia network and pins it to the indexer.
   *
   * @param {Buffer|Uint8Array} data - The data to upload.
   * @param {UploadOptions} [options] - Optional upload options.
   * @returns {Promise<PinnedObject>}
   */
  async upload(data, options) {
    return this.#inner.upload(new BufferReader(data), options ?? DEFAULT_UPLOAD_OPTIONS);
  }

  /**
   * Downloads an object and returns its contents as a Buffer.
   *
   * @param {PinnedObject} obj - The object to download.
   * @param {DownloadOptions} [options] - Optional download options.
   * @returns {Promise<Buffer>}
   */
  async download(obj, options) {
    const w = new BufferWriter();
    await this.#inner.download(w, obj, options ?? DEFAULT_DOWNLOAD_OPTIONS);
    return w.toBuffer();
  }

  /**
   * Creates a new packed upload for efficient multi-object uploads.
   *
   * @param {UploadOptions} [options] - Optional upload options.
   * @returns {Promise<PackedUpload>}
   */
  async upload_packed(options) {
    return new PackedUpload(await this.#inner.upload_packed(options ?? DEFAULT_UPLOAD_OPTIONS));
  }

  app_key() { return this.#inner.app_key(); }
  account() { return this.#inner.account(); }
  delete_object(key) { return this.#inner.delete_object(key); }
  hosts() { return this.#inner.hosts(); }
  object(key) { return this.#inner.object(key); }
  object_events(cursor, limit) { return this.#inner.object_events(cursor, limit); }
  pin_object(obj) { return this.#inner.pin_object(obj); }
  prune_slabs() { return this.#inner.prune_slabs(); }
  share_object(obj, valid_until) { return this.#inner.share_object(obj, valid_until); }
  shared_object(shared_url) { return this.#inner.shared_object(shared_url); }
  slab(slab_id) { return this.#inner.slab(slab_id); }
  update_object_metadata(obj) { return this.#inner.update_object_metadata(obj); }
}

export class PackedUpload {
  #inner;

  /** @internal Use SDK.upload_packed() */
  constructor(inner) {
    this.#inner = inner;
  }

  /**
   * Adds a new object to the upload.
   *
   * @param {Buffer|Uint8Array} data - The data to add.
   * @returns {Promise<bigint|number>} The number of bytes read.
   */
  async add(data) {
    return this.#inner.add(new BufferReader(data));
  }

  finalize() { return this.#inner.finalize(); }
  cancel() { return this.#inner.cancel(); }
  remaining() { return this.#inner.remaining(); }
  length() { return this.#inner.length(); }
  slabs() { return this.#inner.slabs(); }
}

File: python/sia_storage/wrappers.py
"""Idiomatic Python wrappers for the Sia SDK.

This module provides convenience classes that make the SDK feel native to
Python developers: file-like reader/writer support for upload/download,
async iteration and context-manager semantics for streaming downloads, and
plain-callable progress callbacks.
"""

from __future__ import annotations

import asyncio
from io import BytesIO
from typing import Any, BinaryIO, Callable, Optional, Union

from .sia_storage.sia_storage_ffi import (
    AppKey,
    AppMeta,
    Builder as _Builder,
    Download as _Download,
    DownloadOptions,
    PackedUpload as _PackedUpload,
    PinnedObject,
    ProgressCallback,
    Reader,
    Sdk as _Sdk,
    ShardProgress,
    UploadOptions,
    uniffi_set_event_loop,
)


ProgressHandler = Union[ProgressCallback, Callable[[ShardProgress], Any]]


class _CallableProgress(ProgressCallback):
    """Adapts a plain callable into a ProgressCallback trait implementation."""

    def __init__(self, fn: Callable[[ShardProgress], Any]):
        self._fn = fn

    def progress(self, progress: ShardProgress) -> None:
        self._fn(progress)


def _wrap_progress(cb):
    if cb is None or isinstance(cb, ProgressCallback):
        return cb
    if callable(cb):
        return _CallableProgress(cb)
    raise TypeError(
        f"progress callback must be a ProgressCallback or callable, got {type(cb).__name__}"
    )


def _prepare_upload_options(options: Optional[UploadOptions]) -> UploadOptions:
    if options is None:
        return UploadOptions()
    options.shard_uploaded = _wrap_progress(options.shard_uploaded)
    return options


def _prepare_download_options(options: Optional[DownloadOptions]) -> DownloadOptions:
    if options is None:
        return DownloadOptions()
    options.shard_downloaded = _wrap_progress(options.shard_downloaded)
    return options


class Builder(_Builder):
    """Creates a new SDK builder with the provided indexer URL.

    After creating the builder, call connected() to attempt to connect
    using an existing app key, or request_connection() to request a new
    connection.
    """

    def __init__(self, indexer_url: str, app_meta: AppMeta):
        try:
            uniffi_set_event_loop(asyncio.get_running_loop())
        except RuntimeError:
            pass  # No running loop yet
        super().__init__(indexer_url, app_meta)

    async def connected(self, app_key: AppKey) -> Optional[Sdk]:
        """Attempts to connect using the provided app key.

        If the app key is valid, returns an Sdk instance, otherwise returns None.
        If you receive None, call request_connection() to request a new connection.

        Args:
            app_key: The application key used for authentication.
        """
        result = await super().connected(app_key)
        if result is not None:
            return Sdk._from_ffi(result)
        return None

    async def register(self, mnemonic: str) -> Sdk:
        """Registers the application with the indexer using the provided mnemonic.

        Once registered, returns an Sdk instance that can be used to interact
        with the indexer.

        Args:
            mnemonic: The user's mnemonic phrase used to derive the application key.
        """
        return Sdk._from_ffi(await super().register(mnemonic))


class Sdk(_Sdk):
    """The main SDK for interacting with the Sia decentralized storage network.

    Accepts standard Python file-like objects (BinaryIO) for uploads, returns
    an async-iterable Download handle for downloads, and accepts plain Python
    callables for progress reporting.
    """

    def __init__(self, *args, **kwargs):
        raise ValueError("Use Builder.connected() or Builder.register() to create an SDK instance")

    @classmethod
    def _from_ffi(cls, inner: _Sdk):
        # exploits UniFFI's internals to wrap the class
        return cls._uniffi_make_instance(inner._uniffi_clone_handle())

    async def upload(
        self,
        obj: PinnedObject,
        r: Union[bytes, bytearray, memoryview, BinaryIO],
        options: Optional[UploadOptions] = None,
    ) -> PinnedObject:
        """Uploads data to the Sia network.

        Pass `PinnedObject()` for a new upload. To resume a previous upload,
        pass the object returned from the earlier call. Appending data changes
        the object's ID, so any existing references must be updated and the
        object must be re-pinned afterward.

        Args:
            obj: The object to upload into. Use `PinnedObject()` for a new upload.
            r: The data to upload. Accepts a `bytes`-like value or any file-like
                object with a `read(n)` method.
            options: The upload options. `shard_uploaded` accepts either a
                ProgressCallback or any callable taking a ShardProgress.

        Returns:
            An object containing all slabs from `obj` plus the newly uploaded
            slabs. The caller is responsible for pinning the returned object.
        """
        if isinstance(r, (bytes, bytearray, memoryview)):
            r = BytesIO(bytes(r))
        return await super().upload(
            obj,
            BytesReader(r),
            _prepare_upload_options(options),
        )

    async def upload_packed(self, options: Optional[UploadOptions] = None) -> PackedUpload:
        """Creates a new packed upload.

        This allows multiple objects to be packed together for more efficient
        uploads. The returned PackedUpload can be used to add objects to the
        upload, and then finalized to get the resulting objects.

        Args:
            options: The upload options. `shard_uploaded` accepts either a
                ProgressCallback or any callable taking a ShardProgress.

        Returns:
            A PackedUpload that can be used to add objects and finalize the upload.
        """
        return PackedUpload._from_ffi(
            await super().upload_packed(_prepare_upload_options(options))
        )

    def download(self, obj: PinnedObject, options: Optional[DownloadOptions] = None) -> Download:
        """Starts a download of the data referenced by the object.

        Returns a Download handle that streams chunks of decoded data. Use it
        as an async context manager, iterate over it to consume chunks, or
        call read_all() / write_to() for common patterns.

        Args:
            obj: The pinned object to download.
            options: The download options. `shard_downloaded` accepts either a
                ProgressCallback or any callable taking a ShardProgress.

        Returns:
            A Download handle.
        """
        return Download(super().download(obj, _prepare_download_options(options)))


class Download:
    """An async stream of bytes from a downloaded object.

    The handle starts yielding chunks as soon as they become available. Drop
    the handle (via `async with`) or call close() to cancel an in-flight
    download.

    Example:
        # Read into memory
        async with sdk.download(obj) as d:
            data = await d.read_all()

        # Stream chunks to a file
        async with sdk.download(obj) as d:
            with open("out.bin", "wb") as f:
                await d.write_to(f)

        # Iterate over chunks
        async with sdk.download(obj) as d:
            async for chunk in d:
                process(chunk)
    """

    def __init__(self, inner: _Download):
        self._inner = inner

    async def __aenter__(self) -> Download:
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.close()

    def __aiter__(self) -> Download:
        return self

    async def __anext__(self) -> bytes:
        chunk = await self._inner.read()
        if not chunk:
            raise StopAsyncIteration
        return chunk

    async def read(self) -> bytes:
        """Reads the next chunk of decoded data. Returns b'' on EOF."""
        return await self._inner.read()

    async def read_all(self) -> bytes:
        """Reads the entire remaining stream into memory and returns it."""
        parts = []
        while True:
            chunk = await self._inner.read()
            if not chunk:
                break
            parts.append(chunk)
        return b"".join(parts)

    async def write_to(self, w: BinaryIO) -> int:
        """Streams the remaining data to a file-like writer.

        Args:
            w: Any object with a write(data) method.

        Returns:
            The total number of bytes written.
        """
        total = 0
        while True:
            chunk = await self._inner.read()
            if not chunk:
                break
            w.write(chunk)
            total += len(chunk)
        return total

    async def close(self) -> None:
        """Cancels the download and releases any in-flight recovery tasks."""
        await self._inner.cancel()


class PackedUpload(_PackedUpload):
    """A packed upload allows multiple objects to be uploaded together in a single upload.

    This can be more efficient than uploading each object separately if the size
    of the object is less than the minimum slab size.
    """

    def __init__(self, *args, **kwargs):
        raise ValueError("Use SDK.upload_packed() to create a PackedUpload instance")

    @classmethod
    def _from_ffi(cls, inner: _PackedUpload):
        return cls._uniffi_make_instance(inner._uniffi_clone_handle())

    async def add(self, reader: Union[bytes, bytearray, memoryview, BinaryIO]) -> int:
        """Adds a new object to the upload.

        The data will be read until EOF and packed into the upload. The caller
        must call finalize() to get the resulting objects after all objects have
        been added.

        Args:
            reader: The data to add. Accepts a `bytes`-like value or any file-like
                object with a `read(n)` method.

        Returns:
            The number of bytes read.
        """
        if isinstance(reader, (bytes, bytearray, memoryview)):
            reader = BytesIO(bytes(reader))
        return await super().add(BytesReader(reader))


class BytesReader(Reader):
    """Adapts any file-like object to the Reader trait.

    This allows you to upload data from any source that supports the
    standard Python binary read interface (files, BytesIO, etc.).

    Example:
        # From bytes
        reader = BytesReader(BytesIO(b"hello"))

        # From file
        with open("data.bin", "rb") as f:
            reader = BytesReader(f)
    """

    def __init__(self, stream: BinaryIO, chunk_size: int = (1 << 20)):
        """Create a BytesReader from a file-like object.

        Args:
            stream: Any object with a read(n) method returning bytes.
            chunk_size: Size of chunks to read (default 1MiB).
        """
        self._stream = stream
        self._chunk_size = chunk_size

    async def read(self) -> bytes:
        """Read the next chunk of data."""
        return self._stream.read(self._chunk_size)
